File: examples/cpp/runner.cpp
#include <cerrno>
#include <cinttypes>
#include <cstdio>
#include <cstring>
#include <cstdlib>
#include <iostream>
#include <unistd.h>

#include "pins.hpp"
#include "runner.hpp"
#include "pin_utils.hpp"
#include "configuration.hpp"

using std::cerr;

namespace {

constexpr uint8_t MESSAGE_TYPE_ERROR = 0;
constexpr uint8_t MESSAGE_TYPE_PINS = 2;
constexpr size_t PAYLOAD_SIZE = 5;
constexpr size_t MESSAGE_SIZE = PAYLOAD_SIZE + 2; // type + payload + checksum

uint8_t compute_checksum(uint8_t type, const uint8_t* payload)
{
    uint8_t checksum = type;
    for (size_t i = 0; i < PAYLOAD_SIZE; ++i) {
        checksum = static_cast<uint8_t>(checksum + payload[i]);
    }
    return checksum;
}

void read_exact(int fd, uint8_t* buffer, size_t length)
{
    size_t offset = 0;
    while (offset < length) {
        ssize_t bytes = read(fd, buffer + offset, length - offset);
        if (bytes > 0) {
            offset += static_cast<size_t>(bytes);
            continue;
        }
        if (bytes == 0) {
            cerr << "Bridge disconnected while reading" << std::endl;
            std::exit(1);
        }
        if (errno == EINTR) {
            continue;
        }
        cerr << "Read error: " << std::strerror(errno) << std::endl;
        std::exit(1);
    }
}

void write_exact(int fd, const uint8_t* buffer, size_t length)
{
    size_t offset = 0;
    while (offset < length) {
        ssize_t bytes = write(fd, buffer + offset, length - offset);
        if (bytes > 0) {
            offset += static_cast<size_t>(bytes);
            continue;
        }
        if (bytes == 0) {
            cerr << "Bridge disconnected while writing" << std::endl;
            std::exit(1);
        }
        if (errno == EINTR) {
            continue;
        }
        cerr << "Write error: " << std::strerror(errno) << std::endl;
        std::exit(1);
    }
}

} // namespace

Runner::Runner(int device, Memory *mem, bool log_halfcycles)
{
    this->device = device;
    this->mem = mem;
    this->phase = false;
    this->cycle = 0;
    this->log_halfcycles = log_halfcycles;
}

/**
 * CPU reset procedure. It requires to keep the CPU reset pin (40)
 * low for two cycles. After that this pin is set high.
 */
void Runner::reset()
{
    uint8_t b[] = {
        0b00001010, // BE, RW high
        0b00000000,
        0b00000000,
        0b00000000,
        0b10101000  // IRQ, NMI, VDD high
    };

    for (int i = 0; i < 4; i++) {
        this->pins = W65C02Pins(b);
        this->pins.set_overflow = true;
        this->pins.phi2 = this->phase;
        this->write_serial();
        usleep(CYCLE_DURATION);
        this->read_serial();
        this->advance_cycles();
    }

    this->pins.data = 0;
    this->pins.addr = 0;
    this->pins.reset = true;
    this->pins.ready = true;
    this->pins.set_overflow = true;
}

bool Runner::step()
{
    static uint16_t addr = 0;
    static bool write = false;

    this->pins.phi2 = this->phase;
    this->write_serial();
    usleep(CYCLE_DURATION);

    this->read_serial();
    if (!this->phase) {
        addr = this->pins.addr;
        write = this->pins.is_write();
        if (!write) {
            this->pins.data = this->mem->read_byte(addr);
        }
        if (this->log_halfcycles) {
            this->print_state();
        }
    } else {
        uint8_t data = this->pins.data;
        if (write) {
            this->mem->write_byte(addr, data);
        }
        this->print_state();
        if (EXIT_ON_BRK && data == 0 && this->pins.sync) {
            return false;
        }
    }

    this->advance_cycles();
    return MAX_CYCLES == 0 || this->cycle < MAX_CYCLES;
}

void Runner::run()
{
    this->reset();
    while (this->step());
}

void Runner::print_state()
{
    const char rw_char = this->pins.rw ? 'R' : 'W';
    const char phase_char = this->phase ? 'H' : 'L';

    std::printf(
        "Cycle=%06" PRIu64 " Half=%c Addr=$%04X Data=$%02X RW=%c SYNC=%d VP=%d IRQ=%d NMI=%d RES=%d",
        this->cycle,
        phase_char,
        static_cast<unsigned>(this->pins.addr),
        static_cast<unsigned>(this->pins.data),
        rw_char,
        static_cast<int>(this->pins.sync),
        static_cast<int>(this->pins.vector_pull),
        static_cast<int>(this->pins.irq),
        static_cast<int>(this->pins.nmi),
        static_cast<int>(this->pins.reset)
    );

    if (this->log_halfcycles) {
        std::printf(
            " PHI1O=%d PHI2O=%d",
            static_cast<int>(this->pins.phi1o),
            static_cast<int>(this->pins.phi2o)
        );
    }

    std::printf("\n");
}

void Runner::read_serial()
{
    uint8_t type;
    read_exact(this->device, &type, 1);

    if (type == MESSAGE_TYPE_ERROR) {
        uint8_t rest[2]; // error code + checksum
        read_exact(this->device, rest, 2);
        cerr << "Bridge rejected the message with error code "
             << static_cast<int>(rest[0]) << std::endl;
        std::exit(1);
    }

    if (type != MESSAGE_TYPE_PINS) {
        cerr << "Unexpected message type " << static_cast<int>(type) << std::endl;
        std::exit(1);
    }

    uint8_t message[MESSAGE_SIZE];
    message[0] = type;
    read_exact(this->device, message + 1, MESSAGE_SIZE - 1);

    const uint8_t checksum = compute_checksum(message[0], message + 1);
    if (checksum != message[MESSAGE_SIZE - 1]) {
           cerr << "Checksum mismatch: expected " << static_cast<int>(checksum)
               << " got " << static_cast<int>(message[MESSAGE_SIZE - 1]) << std::endl;
        std::exit(1);
    }

    this->pins.set_pins(message + 1);
}

void Runner::write_serial()
{
    uint8_t payload[PAYLOAD_SIZE] = { 0 };
    this->pins.set_buff(payload);

    uint8_t message[MESSAGE_SIZE];
    message[0] = MESSAGE_TYPE_PINS;
    std::memcpy(message + 1, payload, PAYLOAD_SIZE);
    message[MESSAGE_SIZE - 1] = compute_checksum(message[0], payload);

    write_exact(this->device, message, MESSAGE_SIZE);
}

void Runner::advance_cycles()
{
    if (this->phase) {
        ++this->cycle;
    }
    this->phase = !this->phase;
}

File: examples/rust/src/runner.rs
use crate::{
    configuration::{CYCLE_DURATION, SHOW_RAW_DATA},
    pins::Pins,
    protocol::PinsMsg,
};
use serialport::SerialPort;
use std::{
    fs::File,
    io::{self, Read, Write},
    path::PathBuf,
    thread::sleep,
};

pub struct Runner {
    pub cycle: u64,
    pub mem: [u8; 1 << 16],
    pub port: Box<dyn SerialPort>,
    pub phase: bool,
    pub addr: u16,
    pub write: bool,
    pub pins: Pins,
    pub show_halfcycles: bool,
}

impl Runner {
    pub fn run(&mut self) {
        self.reset();
        while self.step() {}
    }
    /// It executes a single half-step (high or low clock signal) of the CPU.
    /// For every exection it sends data to CPU, adjusting the clock status (PHI2) first,
    /// and then it reads status back from the CPU.
    /// For every low half-cycle it reads the address and the RWB status (pin 34).
    /// For every high half-cycle it does the actual memory operation.
    /// At the end the method advances the clock (switches the phase and increments the
    /// clock counter).
    /// The method returns boolean value indicating whether the program reached an end,
    /// that is assumed to be represented by BRK instruction (opcode 0).
    pub fn step(&mut self) -> bool {
        self.pins.phi2 = self.phase;
        self.write_port();

        sleep(CYCLE_DURATION);
        self.read_port();

        if !self.phase {
            self.addr = self.pins.addr;
            self.write = self.pins.is_write();
            if !self.write {
                self.pins.data = read_byte(&self.mem, self.addr);
            }
        } else {
            if self.write {
                write_byte(&mut self.mem, self.addr, self.pins.data);
            }
            self.print_state();
            if self.pins.sync && self.pins.data == 0 && self.cycle > 20 {
                return false;
            }
        }

        if !self.phase {
            if self.show_halfcycles {
                self.print_state();
            }
        }

        self.advance_cycles();
        true
    }

    /// Resets the CPU, by holding RESB signal (pin 40) low for two cycles (4 half-cycles).
    pub fn reset(&mut self) {
        let mut pins = Pins::default();
        pins.be = true;
        pins.rw = true;
        pins.irq = true;
        pins.nmi = true;
        pins.so = false;
        pins.reset = false;

        for _ in 0..4 {
            self.pins = pins;
            self.pins.phi2 = self.phase;
            self.write_port();

            sleep(CYCLE_DURATION);
            self.read_port();
            self.advance_cycles();
        }

        self.pins.data = 0;
        self.pins.addr = 0;
        self.pins.reset = true;
        self.pins.ready = true;
        self.pins.so = true;
    }

    fn print_state(&self) {
        let rw_char = if self.pins.rw { 'R' } else { 'W' };
        let phase_char = if self.phase { 'H' } else { 'L' };
        print!(
            "Cycle={cycle:06} Half={phase} Addr=${addr:04X} Data=${data:02X} RW={rw} SYNC={sync} VP={vp} IRQ={irq} NMI={nmi} RES={res}",
            cycle = self.cycle,
            phase = phase_char,
            addr = self.pins.addr,
            data = self.pins.data,
            rw = rw_char,
            sync = u8::from(self.pins.sync),
            vp = u8::from(self.pins.vp),
            irq = u8::from(self.pins.irq),
            nmi = u8::from(self.pins.nmi),
            res = u8::from(self.pins.reset),
        );

        if self.show_halfcycles {
            print!(
                " PHI1O={} PHI2O={}",
                u8::from(self.pins.phi1o),
                u8::from(self.pins.phi2o)
            );
        }

        println!();
    }

    /// Reads a message from the serial port and updates `pins` field.
    /// The bridge responds either with a 7-byte pins message (type 2),
    /// or with a 3-byte error message (type 0) when it rejects a request.
    fn read_port(&mut self) {
        let mut header = [0u8; 1];
        self.port
            .read_exact(&mut header)
            .expect("Read error from serial port");

        match header[0] {
            0 => {
                let mut rest = [0u8; 2]; // error code + checksum
                self.port
                    .read_exact(&mut rest)
                    .expect("Read error from serial port");
                panic!("Bridge rejected the message with error code {}", rest[0]);
            }
            2 => {
                let mut buff = [0u8; 7];
                buff[0] = header[0];
                self.port
                    .read_exact(&mut buff[1..])
                    .expect("Read error from serial port");
                let msg = PinsMsg::from_bytes(&buff[..]);
                if SHOW_RAW_DATA {
                    println!("Reading: {}", msg);
                    // print_buff(&buff);
                }
                self.pins = Pins::from(msg.data);
            }
            t => panic!("Unexpected message type: {}", t),
        }
    }

    /// Send the value of `pins` field into to serial port.
    fn write_port(&mut self) {
        let payload: [u8; 5] = self.pins.into();
        let msg = PinsMsg::new(2, payload);
        if SHOW_RAW_DATA {
            print!("Writing: {}", msg);
            // print_buff(&buff);
        }
        self.port
            .write_all(&msg.to_vec())
            .expect("Write error to serial port");
    }

    /// Changes phase of the clock and advances the clock count.
    fn advance_cycles(&mut self) {
        if self.phase {
            self.cycle += 1;
        }
        self.phase = !self.phase;
    }
}

/// Writes a byte under a given address to memory emulation.
pub fn write_byte(mem: &mut [u8], addr: u16, val: u8) {
    mem[addr as usize] = val;
}

/// Reads a byte from a given address of memeory emulation.
pub fn read_byte(mem: &[u8], addr: u16) -> u8 {
    mem[addr as usize]
}

/// Reads contnet of a 6502 program and returns it as
/// a vector of bytes.
fn get_file_as_byte_vec(filename: &PathBuf) -> io::Result<Vec<u8>> {
    let mut f = File::open(filename)?;
    let mut buffer = Vec::new();
    f.read_to_end(&mut buffer)?;
    Ok(buffer)
}

/// Loads 6502 program into memory emulation under a given address.
pub fn load_program(mem: &mut [u8], file: &str, addr: u16) {
    let prg = get_file_as_byte_vec(&PathBuf::from(file)).expect("Can't open program file");
    for (i, byte) in prg.iter().enumerate() {
        write_byte(mem, addr + (i as u16), *byte);
    }
}

/// Prints serial port buffer in a form of 4x8 bits.
fn print_buff(buff: &[u8]) {
    for i in 0..buff.len() {
        print!("{:08b} ", buff[i]);
    }
    println!();
}
